python_modules/dagster/dagster/_core/definitions/asset_checks.py

```diff
@@ -127,3 +143,75 @@ def get_attributes_dict(self) -> Dict[str, Any]:
             specs=self._specs,
             input_output_props=self._input_output_props,
         )
+
+
+@experimental
+def build_blocking_asset_check(
+    asset_def: "AssetsDefinition",
+    checks: Sequence[AssetChecksDefinition],
+) -> "AssetsDefinition":
+    from dagster import AssetIn, In, OpExecutionContext, Output, op
+    from dagster._core.definitions.decorators.asset_decorator import graph_asset_no_defaults
+    from dagster._core.storage.asset_check_execution_record import AssetCheckExecutionRecordStatus
+
+    check_specs = []
+    for c in checks:
+        check_specs.extend(c.specs)
+
+    check_output_names = [c.get_python_identifier() for c in check_specs]
+
+    check.invariant(len(asset_def.op.output_defs) == 1)
+    asset_out_type = asset_def.op.output_defs[0].dagster_type
+
+    @op(ins={"asset_result": In(asset_out_type), "check_evaluations": In(Nothing)})
+    def fan_in_checks_and_asset_result(context: OpExecutionContext, asset_result):
```

Comment:
are we not type-hinting `asset_result` on purpose because of dagster typing machinery?

Reply:
https://github.com/dagster-io/dagster/pull/16612#discussion_r1336468118

what would the hint be? The input is the output from the @asset decorated function, I don't know how to turn that in to a type variable. You can get it at runtime with `op_def.get_output_annotation()`

Reply:
I think we'd have to do some very fancy generics with @asset down to the AssetsDefinition to here to get an annotation

Reply:
Got it understood. Just typing `asset_result` as `Any` would be helpful. Using the term "result" is a little problematic here since we now have "Result" objects and this was the source of my confusion. I think choosing a different variable name would be helpful. `asset_return_value` perhaps.